modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java

```diff
@@ -970,28 +976,41 @@ private CompletableFuture<Void> continueIndexScan(
                         return completedFuture(null); // End of range reached. Exit loop.
                     }
 
-                    return lockManager.acquire(txId, new LockKey(tableId(), currentRow.rowId()), LockMode.S)
+                    RowId rowId = currentRow.rowId();
+
+                    return lockManager.acquire(txId, new LockKey(tableId(), rowId), LockMode.S)
                             .thenComposeAsync(rowLock -> { // Table row S lock
-                                ReadResult readResult = mvDataStorage.read(currentRow.rowId(), HybridTimestamp.MAX_VALUE);
-                                return resolveAndCheckReadCompatibility(readResult, txId)
-                                        .thenCompose(resolvedReadResult -> {
-                                            if (resolvedReadResult != null) {
-                                                if (indexRowMatches(currentRow, resolvedReadResult, schemaAwareIndexStorage)) {
-                                                    result.add(resolvedReadResult);
-                                                }
-                                            }
-
-                                            // Proceed scan.
-                                            return continueIndexScan(
-                                                    txId,
-                                                    schemaAwareIndexStorage,
-                                                    indexLocker,
-                                                    indexCursor,
-                                                    batchSize,
-                                                    result,
-                                                    isUpperBoundAchieved
-                                            );
-                                        });
+                                ReadResult readResult = mvDataStorage.read(rowId, HybridTimestamp.MAX_VALUE);
+                                return resolveAndCheckReadCompatibility(readResult, txId, () -> {
+                                    if (readResult.newestCommitTimestamp() == null) {
```

Comment:
Why do we have different logic for candidate calculation. Within code above there's a check that given readResult is a writeIntent.
```
            BinaryRow candidate =
                    newestCommitTimestamp == null || !readResult.isWriteIntent() ? null : cursor.committed(newestCommitTimestamp);
```
However here we don't have such logic, why?

Reply:
Because here we can be sure that is a write-intent. It happens after the resolution is finalized and the decision about getting committed value is applied.

Reply:
I am not sure it is needed. We already have the same logic for the resolution in other places. In the code where we check it for a cursor, we are doing this to avoid additional calculations on a commit candidate, but when the closure is invoked, we might be sure the committed value is really needed.

Reply:
So, you believe that multiple code duplications are better?